radix-engine-stores/src/rocks_db.rs

```diff
@@ -16,52 +17,64 @@ impl RocksdbSubstateStore {
 }
 
 impl SubstateDatabase for RocksdbSubstateStore {
-    fn get_substate(&self, index_id: &Vec<u8>, db_key: &Vec<u8>) -> Option<Vec<u8>> {
-        let key = encode_substate_id(index_id, db_key);
-        self.db.get(&key).expect("IO Error")
+    fn get_substate(
+        &self,
+        partition_key: &DbPartitionKey,
+        sort_key: &DbSortKey,
+    ) -> Option<DbSubstateValue> {
+        let key_bytes = encode_to_rocksdb_bytes(partition_key, sort_key);
+        self.db.get(&key_bytes).expect("IO Error")
     }
 
-    fn list_substates(
+    fn list_entries(
         &self,
-        index_id: &Vec<u8>,
-    ) -> Box<dyn Iterator<Item = (Vec<u8>, Vec<u8>)> + '_> {
-        let index_id = index_id.clone();
-
-        let start = encode_substate_id(&index_id, &vec![0]);
-
+        partition_key: &DbPartitionKey,
+    ) -> Box<dyn Iterator<Item = PartitionEntry> + '_> {
+        let partition_key = partition_key.clone();
+        let start_key_bytes = encode_to_rocksdb_bytes(&partition_key, &DbSortKey(vec![]));
         let iter = self
             .db
-            .iterator(IteratorMode::From(&start, Direction::Forward))
-            .take_while(move |kv| {
-                let (key, _value) = kv.as_ref().unwrap();
-                key[0..26].eq(&index_id)
-            })
+            .iterator(IteratorMode::From(&start_key_bytes, Direction::Forward))
             .map(|kv| {
-                let (key, value) = kv.unwrap();
-                let (_, substate_key) =
-                    decode_substate_id(key.as_ref()).expect("Failed to decode substate ID");
-                let value = value.as_ref().to_vec();
-                (substate_key, value)
-            });
+                let (iter_key_bytes, iter_value) = kv.as_ref().unwrap();
+                let iter_key = decode_from_rocksdb_bytes(iter_key_bytes);
+                (iter_key, iter_value.to_vec())
+            })
+            .take_while(move |((iter_partition_key, _), _)| *iter_partition_key == partition_key)
```

Comment:
This is slightly less efficient than the old approach, as we're to vec'ing the full substate in the next DB key - which could potentially be a full package substate - I wonder if we could do the take while first by checking the key prefix, like the old approach?

Reply:
(We have a separate implementation of this in the node anyway, so this is not a big deal)

Reply:
AFAIU, this impl (in scrypto repo) is for tests and local experiments, so let's go for simplicity/readability.

We'll do the performance-driven raw-byte-prefix comparisons in our Node impl, as you say 👍